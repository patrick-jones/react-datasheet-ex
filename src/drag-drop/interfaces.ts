import RDS from 'react-datasheet';
import {ConnectDragPreview, ConnectDragSource, ConnectDropTarget} from 'react-dnd';
import {ColumnHeader, HeaderRendererProps, RowRendererProps} from '../interfaces';

/**
 * Properties that must be provided to a drag and drop enabled row
 */
export interface DragDropRowProps<T extends RDS.Cell<T, V>, V = string> {
  /** The current row index */
  row: number;
  /** Callback for when a row is dropped on another */
  readonly onRowDropped: (source: number, target: number) => void;
}

/**
 * Properties that must be provided to a drag and drop enabled column header
 */
export interface DragDropHeaderProps {
  /** The current column index */
  col: number;
  /** Callback for one header is dropped on another */
  readonly onHeaderDropped: (source: number, target: number) => void;
}

/**
 * Drag source properties that are injected into a drag and drop provider
 */
export interface InjectedDragSourceProps {
  readonly connectDragSource: ConnectDragSource;
  readonly connectDragPreview: ConnectDragPreview;
}

/**
 * Drop target properties that are injected into a drag and drop provider
 */
export interface InjectedDropTargetProps {
  readonly connectDropTarget: ConnectDropTarget;
  readonly isDropOver: boolean;
}

/**
 * Combined drag source and drop target properties that are injected
 * into a drag and drop provider
 */
export interface InjectedDragDropProps extends InjectedDragSourceProps, InjectedDropTargetProps {}

/**
 * Extends `ReactDatasheet.RowRendererProps` to include
 * drag and drop properties
 */
export interface DragDropRowRendererProps<CellType, V = string>
  extends RowRendererProps<CellType, V>,
    DragDropRowProps<CellType, V> {}

/**
 * Extends `HeaderRendererProps` to include drag and drop properties.
 */
export interface DragDropHeaderRendererProps<T extends ColumnHeader = ColumnHeader>
  extends HeaderRendererProps<T>,
    DragDropHeaderProps {}
