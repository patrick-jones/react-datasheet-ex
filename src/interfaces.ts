import React, {CSSProperties} from 'react';
import RDS from 'react-datasheet';

export interface RowLocation {
  row: number;
}

export interface ColumnLocation {
  col: number;
}

export interface ColumnHeader {
  id: string;
  title: string;
  width?: number;
}

export interface HeaderRendererProps<T extends ColumnHeader = ColumnHeader> extends ColumnLocation {
  header: T;
}

export type HeaderRenderer<T extends ColumnHeader = ColumnHeader,
  R extends HeaderRendererProps<T> = HeaderRendererProps<T>> = React.ComponentType<R>;

export type HeaderContentRenderer<T extends ColumnHeader = ColumnHeader>
  = React.ComponentType<HeaderRendererProps<T>>;

export interface SheetRendererProps<
    T extends RDS.Cell<T, V>, V = string, H extends ColumnHeader = ColumnHeader
  >
  extends RDS.SheetRendererProps<T, V> {
  /** React style object that should be applied to the enclosing table */
  style?: CSSProperties | undefined;
  /**
   * An array of `ColumnHeader`, which are rendered into `th` elements
   * @default []
   */
  headers: H[];
  /**
   * Optional component or function to render the header cell contents.
   * The contents are rendered within a `th` element.
   * @default DefaultHeaderRenderer
   */
  headerRenderer: HeaderRenderer;
  /**
   * Optional component or function to render the header cell contents.
   * The contents are rendered within a `th` element.
   * @default DefaultHeaderContentRenderer
   */
  headerContentRenderer: HeaderContentRenderer;
}


export type ActionContentRenderer<T extends RowLocation = RowLocation>
  = React.ComponentType<T>;

export interface RowRendererProps<
    T extends RDS.Cell<T, V>, V = string, AP extends RowLocation = RowLocation
  > extends RDS.RowRendererProps<T, V> {

  actionContentRenderer?: ActionContentRenderer<AP>;
}
