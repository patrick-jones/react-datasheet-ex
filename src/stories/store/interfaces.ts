import RDS from 'react-datasheet';
import {ColumnHeader} from '../../interfaces';

export type ExampleCellValue = string | number | null;
export interface ExampleCellType extends RDS.Cell<ExampleCellType> {
  value: ExampleCellValue;
}

export type ExampleModel = {
  id: string;
  headers: ColumnHeader[];
  rows: ExampleRow[];
  selected: RDS.Selection | null;
};

export type ExampleRow = {
  id: string;
  selected?: boolean;
  data: ExampleCellType[];
};

export type Grid = ExampleCellType[][];
