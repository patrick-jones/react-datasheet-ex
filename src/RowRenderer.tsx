import React, {Component, ReactNode} from 'react';
import ReactDatasheet from 'react-datasheet';

import {RowLocation, RowRendererProps} from './interfaces';
import DefaultActionContentRenderer from './DefaultActionContentRenderer';


export default class RowRenderer<
    T extends ReactDatasheet.Cell<T, V>, V = string, AP extends RowLocation = RowLocation
  > extends Component<RowRendererProps<T, V, AP>> {

  static displayName = 'RowRenderer';

  static readonly defaultProps = {
    actionContentRenderer: DefaultActionContentRenderer,
  };

  render(): ReactNode {
    const {row, children, actionContentRenderer} = this.props;
    const ContentRenderer = actionContentRenderer || DefaultActionContentRenderer;
    return (
      <tr key={`$row-${row}`}>
        <td key='$$actionCell' className='cell read-only rdr-sheet-renderer__action-cell'>
          <ContentRenderer row={row} />
        </td>
        {children}
      </tr>
    );
  }
}
