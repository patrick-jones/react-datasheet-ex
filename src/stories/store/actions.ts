import RDS from 'react-datasheet';
import {ExampleCellType} from './interfaces';

/**
 * Some quick redux-like interfaces
 */

/**
 * A redux-like action
 */
export type Action<P = any> = {
  readonly type: string;
  readonly payload: P;
};

/**
 * Abstract type for a function that creates an action
 */
export type ActionCreator = (...args: any[]) => Action;

/**
 * Abstract type for a mapping of keys are ActionCreator functions
 */
export type ActionsCreators = {
  [key: string]: ActionCreator;
};

/**
 * Abstract type for the Action objects returned by an ActionCreators mapping
 */
export type Actions<A extends ActionsCreators> = ReturnType<A[keyof A]>;

/**
 * Type for actions once they are run through the dispatch function
 */
export type DispatchedActionCreator<AC extends ActionCreator> = (...args: any[]) => void;

/**
 * Converts `actions` object from functions that return an Action to
 * functions with the same arguments that return void.
 */
export type DispatchedActions<TAC> = {
  [K in keyof TAC]: TAC[K] extends ((...args: infer A) => Action) ? (...args: A) => void : undefined
};

const createAction = <T extends string, P>(type: T, payload: P) => ({type, payload});

/**
 * Our specific actions
 */
export enum ActionTypes {
  CELLS_CHANGED = 'rds-ex__cells-changed',
  SELECTION_CHANGED = 'rds-ex__selection-changed',
  HEADER_DROPPED = 'rds-ex__header-dropped',
  ROW_DROPPED = 'rds-ex__row-dropped',
}

export const SheetActionCreators = {
  cellsChanged:
    (changes: RDS.CellsChangedArgs<ExampleCellType>, additions?: RDS.CellAdditionsArgs<string>) =>
      createAction(ActionTypes.CELLS_CHANGED, {changes, additions}),
  selectionChanged: (selected: RDS.Selection) =>
    createAction(ActionTypes.SELECTION_CHANGED, {selected}),
  headerDropped: (source: number, target: number) =>
    createAction(ActionTypes.HEADER_DROPPED, {source, target}),
  rowDropped: (source: number, target: number) =>
    createAction(ActionTypes.ROW_DROPPED, {source, target}),
};

export type SheetActionCreators = typeof SheetActionCreators;
