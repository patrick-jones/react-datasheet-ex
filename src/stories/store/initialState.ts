import {ExampleModel} from './interfaces';

export const initialState: ExampleModel = {
  id: 'foobar',
  headers: [
    {id: 'A', title: 'A'},
    {id: 'B', title: 'B'},
    {id: 'C', title: 'C'},
    {id: 'D', title: 'D'},
  ],
  rows: [
    {id: '1', data: [{value: 1}, {value: 2}, {value: 3}, {value: 4}]},
    {id: '2', data: [{value: 5}, {value: 6}, {value: 7}, {value: 8}]},
    {id: '3', data: [{value: 9}, {value: 10}, {value: 11}, {value: 12}]},
  ],
  selected: null,
};
