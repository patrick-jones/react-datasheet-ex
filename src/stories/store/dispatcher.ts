import {Store} from '@dump247/storybook-state';
import {ExampleModel} from './interfaces';
import {Actions, ActionsCreators, DispatchedActionCreator, DispatchedActions} from './actions';
import reducer from './reducer';
import {action} from '@storybook/addon-actions';

function dispatch<AC extends ActionsCreators>(store: Store<ExampleModel>, a: Actions<AC>) {
  action(a.type)(a.payload);
  const nextState = reducer(store.state, a);
  if (nextState && nextState !== store.state) {
    store.set(nextState);
  }
}

export default function bindDispatch<AC extends ActionsCreators>(
  store: Store<ExampleModel>,
  actionCreators: AC
): DispatchedActions<AC> {
  const bound = {} as any;
  Object.keys(actionCreators).forEach(key => {
    const orig = actionCreators[key];
    const ds: DispatchedActionCreator<typeof orig> = (...args) =>
      dispatch(store, actionCreators[key](...args));
    bound[key] = ds;
  });
  return bound as DispatchedActions<AC>;
}
