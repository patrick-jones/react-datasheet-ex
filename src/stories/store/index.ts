import {initialState} from './initialState';

export {initialState};
