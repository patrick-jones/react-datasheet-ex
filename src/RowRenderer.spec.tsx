import React from 'react';
import {mount} from 'enzyme';
import toJson from 'enzyme-to-json';

import RowRenderer from './RowRenderer';
import {ExampleCellType} from './stories/store/interfaces';

const cells: ExampleCellType[] = [
  {value: 'cell'},
];

describe('RowRenderer basic rendering', () => {
  it('Should render', () => {
    const wrapper = mount(
      <table>
        <tbody>
          <RowRenderer row={0} cells={cells}>
            <td className='cell'>cell</td>
          </RowRenderer>
        </tbody>
      </table>
    );

    const td = wrapper.find('td.cell');
    expect(td.length).toBe(2);
    expect(td.find('td.rdr-sheet-renderer__action-cell').length).toBe(1);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('Should render custom action content', () => {
    const wrapper = mount(
      <table>
        <tbody>
        <RowRenderer
          row={0}
          cells={cells}
          actionContentRenderer={() => (<span className='abcd-puppy'>Action</span>)}
        >
          <td className='cell'>cell</td>
        </RowRenderer>
        </tbody>
      </table>
    );

    expect(wrapper.find('td.rdr-sheet-renderer__action-cell span.abcd-puppy').text())
      .toBe('Action');
  });
});
