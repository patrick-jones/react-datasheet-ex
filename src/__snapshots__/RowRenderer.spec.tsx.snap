// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RowRenderer basic rendering Should render 1`] = `
<table>
  <tbody>
    <RowRenderer
      actionContentRenderer={[Function]}
      cells={
        Array [
          Object {
            "value": "cell",
          },
        ]
      }
      row={0}
    >
      <tr
        key="$row-0"
      >
        <td
          className="cell read-only rdr-sheet-renderer__action-cell"
          key="$$actionCell"
        >
          <DefaultActionContentRenderer
            row={0}
          />
        </td>
        <td
          className="cell"
        >
          cell
        </td>
      </tr>
    </RowRenderer>
  </tbody>
</table>
`;
